File: MasterProject/Code/analysis.R
# ============================================================
# VecTraits Heterogeneity Analysis & Semantic Clustering
# ============================================================
Sys.setenv(RETICULATE_PYTHON = "/Users/danielzhu/Library/r-miniconda-arm64/envs/pymc-env/bin/python")
library(reticulate)
library(tidyverse)



setwd("/Users/danielzhu/Documents/Master Project")

# ── 1. read data ───────────────────────────────────────────────
rows <- read_csv("Data/VecTraits.csv")

cat("Total rows:", nrow(rows), "\n")

# ── 2. unique trait names ───────────────────────────────────────────────
unique_names <- rows %>%
  mutate(OriginalTraitName = str_trim(OriginalTraitName)) %>%
  count(OriginalTraitName, sort = TRUE)

cat("Unique OriginalTraitNames:", nrow(unique_names), "\n")

# ── 3. Name + Definition paring ─────────────────────────────────
pairs <- rows %>%
  mutate(across(c(OriginalTraitName, OriginalTraitDef), str_trim)) %>%
  count(OriginalTraitName, OriginalTraitDef, sort = TRUE)

cat("Unique name+definition pairs:", nrow(pairs), "\n")

# ── 4. one name -> multiple definitions ───────────────────────────────────
name_to_defs <- rows %>%
  mutate(across(c(OriginalTraitName, OriginalTraitDef), str_trim)) %>%
  group_by(OriginalTraitName) %>%
  summarise(n_definitions = n_distinct(OriginalTraitDef)) %>%
  ungroup()

multi_def <- name_to_defs %>% filter(n_definitions > 1)
cat("Names with multiple definitions:", nrow(multi_def), "\n")

# ── 5. StandardisedTraitName coverage──────────────────────────
std_filled <- rows %>%
  filter(str_trim(StandardisedTraitName) != "" & !is.na(StandardisedTraitName))

cat("StandardisedTraitName filled:", nrow(std_filled), "/", nrow(rows), "\n")
cat(sprintf("(%.1f%%)\n", 100 * nrow(std_filled) / nrow(rows)))

unique_std <- std_filled %>%
  count(StandardisedTraitName, sort = TRUE)

cat("Unique StandardisedTraitName values:", nrow(unique_std), "\n")
cat("Top 5:\n")
print(head(unique_std, 5))

# ── 6. unit consistency ─────────────────────────────────────────────
name_to_units <- rows %>%
  mutate(across(c(OriginalTraitName, OriginalTraitUnit), str_trim)) %>%
  filter(OriginalTraitUnit != "" & !is.na(OriginalTraitUnit)) %>%
  group_by(OriginalTraitName) %>%
  summarise(n_units = n_distinct(OriginalTraitUnit)) %>%
  ungroup()

multi_unit <- name_to_units %>% filter(n_units > 1)
cat("Names with multiple units:", nrow(multi_unit), "\n")

# ── 7. Top 20 most frequent trait names ──────────────────────────────
cat("\nTop 20 most frequent trait names:\n")
top20 <- unique_names %>%
  head(20) %>%
  left_join(name_to_defs, by = "OriginalTraitName") %>%
  left_join(name_to_units, by = "OriginalTraitName") %>%
  replace_na(list(n_definitions = 1, n_units = 0))

print(top20)

# ── 8. Taxa analysis ──────────────────────────────────────────────
name_to_taxa <- rows %>%
  mutate(across(c(OriginalTraitName, Interactor1Genus), str_trim)) %>%
  filter(Interactor1Genus != "" & !is.na(Interactor1Genus)) %>%
  group_by(OriginalTraitName) %>%
  summarise(n_genera = n_distinct(Interactor1Genus)) %>%
  ungroup()

# ── 9. captilisation inconsistency ───────────────────────────────────────
capitalisation_issues <- unique_names %>%
  mutate(name_lower = str_to_lower(OriginalTraitName)) %>%
  group_by(name_lower) %>%
  filter(n() > 1) %>%
  ungroup() %>%
  arrange(name_lower)

cat("\nCapitalisation inconsistencies:\n")
print(capitalisation_issues)

# ── 10. generate heterogeneity analysis CSV ──────────────────────────────────────────
output_table <- unique_names %>%
  rename(total_rows = n) %>%
  left_join(name_to_defs, by = "OriginalTraitName") %>%
  left_join(name_to_units, by = "OriginalTraitName") %>%
  left_join(name_to_taxa, by = "OriginalTraitName") %>%
  replace_na(list(n_definitions = 1, n_units = 0, n_genera = 0)) %>%
  mutate(
    taxa_specific = if_else(n_genera <= 3 & total_rows >= 100, "YES", ""),
    capitalisation_issue = if_else(
      str_to_lower(OriginalTraitName) %in%
        (capitalisation_issues %>% pull(name_lower)),
      "YES", ""
    )
  )

dir.create("Results", showWarnings = FALSE)
write_csv(output_table, "Results/heterogeneity_analysis.csv")
cat("Saved to Results/heterogeneity_analysis.csv\n")

# ── 11.  embedding input ───────────────────────────────
cat("\nPreparing text for embedding...\n")

unique_pairs <- rows %>%
  mutate(across(c(OriginalTraitName, OriginalTraitDef, OriginalTraitUnit),
                str_trim)) %>%
  distinct(OriginalTraitName, OriginalTraitDef, .keep_all = TRUE) %>%
  select(OriginalTraitName, OriginalTraitDef, OriginalTraitUnit) %>%
  mutate(
    text = paste0(
      "Trait: ", OriginalTraitName,
      ". Definition: ", OriginalTraitDef,
      ". Unit: ", OriginalTraitUnit, "."
    )
  )

cat("Total unique pairs to embed:", nrow(unique_pairs), "\n")
cat("Example:", unique_pairs$text[1], "\n")

# ── 12. generate Embeddings (via OpenAI API) ──────────────────────
library(httr)
library(jsonlite)

cat("\nGenerating embeddings with text-embedding-3-large...\n")

Sys.setenv(OPENAI_API_KEY = "")

get_embeddings <- function(texts, model = "text-embedding-3-large") {
  body <- list(
    input = as.list(texts),
    model = model
  )
  
  response <- POST(
    url = "https://api.openai.com/v1/embeddings",
    add_headers(
      "Authorization" = paste("Bearer", Sys.getenv("OPENAI_API_KEY")),
      "Content-Type" = "application/json"
    ),
    body = toJSON(body, auto_unbox = TRUE),
    encode = "raw"
  )
  
  result <- content(response, as = "parsed")
  
  if (!is.null(result$error)) {
    cat("API Error:", result$error$message, "\n")
    return(NULL)
  }
  
  embeddings <- do.call(rbind, lapply(result$data, function(x) unlist(x$embedding)))
  return(embeddings)
}


batch_size <- 100
all_embeddings <- list()

for (i in seq(1, nrow(unique_pairs), by = batch_size)) {
  batch <- unique_pairs$text[i:min(i + batch_size - 1, nrow(unique_pairs))]
  cat(sprintf("Processing batch %d/%d...\n",
      ceiling(i/batch_size),
      ceiling(nrow(unique_pairs)/batch_size)))
  all_embeddings[[length(all_embeddings) + 1]] <- get_embeddings(batch)
}

embeddings <- do.call(rbind, all_embeddings)
cat("Embeddings shape:", dim(embeddings), "\n")


# ── 13. HDBSCAN clustering ──────────────────────────────────────────
cat("\nClustering...\n")

hdbscan_pkg <- import("hdbscan")

clusterer <- hdbscan_pkg$HDBSCAN(
  min_cluster_size = 5L,
  min_samples = 3L,
  metric = "euclidean"
)

cluster_labels <- clusterer$fit_predict(embeddings)

n_clusters <- length(unique(cluster_labels[cluster_labels != -1]))
n_noise <- sum(cluster_labels == -1)

cat("Number of clusters:", n_clusters, "\n")
cat("Noise points (unclustered):", n_noise, "\n")
cat("Clustered points:", length(cluster_labels) - n_noise, "\n")

# ── 14. Save clustering results ──────────────────────────────────────────
results <- unique_pairs %>%
  select(OriginalTraitName, OriginalTraitDef, OriginalTraitUnit, text) %>%
  mutate(
    cluster_id = as.vector(cluster_labels),
    n_members = as.integer(table(as.vector(cluster_labels))[as.character(as.vector(cluster_labels))])
  ) %>%
  arrange(cluster_id)

write_csv(results, "Results/clustering_results_openai.csv")
cat("Saved to Results/clustering_results_openai.csv\n")

sink("Results/clustering_results_openai.txt")
for (cid in sort(unique(results$cluster_id))) {
  if (cid == -1) next
  members <- results %>% filter(cluster_id == cid)
  cat(sprintf("\n--- Cluster %d (%d members) ---\n", cid, nrow(members)))
  for (i in seq_len(nrow(members))) {
    cat(sprintf("  '%s' | %s | %s\n",
                members$OriginalTraitName[i],
                str_trunc(members$OriginalTraitDef[i], 50),
                members$OriginalTraitUnit[i]))
  }
}
noise <- results %>% filter(cluster_id == -1)
cat(sprintf("\n--- NOISE (unclustered, %d items) ---\n", nrow(noise)))
for (i in seq_len(nrow(noise))) {
  cat(sprintf("  '%s' | %s | %s\n",
              noise$OriginalTraitName[i],
              str_trunc(noise$OriginalTraitDef[i], 50),
              noise$OriginalTraitUnit[i]))
}
sink()

cat("Saved to Results/clustering_results_openai.txt\n")
cat("\nAll done!\n")

File: MasterProject/Code/analysis.py
import csv
from collections import defaultdict, Counter


rows = []
with open('../Data/VecTraits.csv', encoding='utf-8') as f:
    reader = csv.DictReader(f)
    for row in reader:
        rows.append(row)


trait_names = [r['OriginalTraitName'].strip() for r in rows]
unique_names = Counter(trait_names)
print(f"Total rows: {len(rows)}")
print(f"Unique OriginalTraitNames: {len(unique_names)}")

pairs = Counter()
for r in rows:
    key = (r['OriginalTraitName'].strip(), r['OriginalTraitDef'].strip())
    pairs[key] += 1
print(f"Unique name+definition pairs: {len(pairs)}")

name_to_defs = defaultdict(set)
for r in rows:
    name_to_defs[r['OriginalTraitName'].strip()].add(r['OriginalTraitDef'].strip())
multi_def = {k: v for k, v in name_to_defs.items() if len(v) > 1}
print(f"Names with multiple definitions: {len(multi_def)}")

std_filled = [r for r in rows if r['StandardisedTraitName'].strip() != '']
print(f"\nStandardisedTraitName filled: {len(std_filled)}/{len(rows)} "
      f"({100*len(std_filled)/len(rows):.1f}%)")
unique_std = Counter([r['StandardisedTraitName'].strip() for r in std_filled])
print(f"Unique StandardisedTraitName values: {len(unique_std)}")
print("Top 5:")
for name, count in unique_std.most_common(5):
    print(f"  [{count:5d}]  {name}")


name_to_units = defaultdict(set)
for r in rows:
    unit = r['OriginalTraitUnit'].strip()
    if unit:
        name_to_units[r['OriginalTraitName'].strip()].add(unit)
multi_unit = {k: v for k, v in name_to_units.items() if len(v) > 1}
print(f"\nNames with multiple units: {len(multi_unit)}")


print("\nTop 20 most frequent trait names:")
for name, count in unique_names.most_common(20):
    n_defs = len(name_to_defs[name])
    n_units = len(name_to_units.get(name, set()))
    print(f"  [{count:5d} rows]  {n_defs} defs  {n_units} units  '{name}'")


name_to_taxa = defaultdict(set)
for r in rows:
    genus = r['Interactor1Genus'].strip()
    if genus:
        name_to_taxa[r['OriginalTraitName'].strip()].add(genus)

print("\nTop 10 trait names and genera coverage:")
for name, count in unique_names.most_common(10):
    n_taxa = len(name_to_taxa.get(name, set()))
    print(f"  [{count:5d} rows]  {n_taxa:3d} genera  '{name}'")

print("\nCapitalisation inconsistencies:")
lower_names = defaultdict(list)
for name in set(trait_names):
    lower_names[name.lower()].append(name)
for lower, variants in lower_names.items():
    if len(variants) > 1:
        counts = [unique_names[v] for v in variants]
        print(f"  {variants}  (rows: {counts})")


import os

output_rows = []
for name, count in unique_names.most_common():
    output_rows.append({
        'OriginalTraitName': name,
        'total_rows': count,
        'n_definitions': len(name_to_defs[name]),
        'n_units': len(name_to_units.get(name, set())),
        'n_genera': len(name_to_taxa.get(name, set())),
        'taxa_specific': 'YES' if len(name_to_taxa.get(name, set())) <= 3 and count >= 100 else '',
        'capitalisation_issue': 'YES' if any(
            name.lower() == other.lower() and name != other
            for other in unique_names.keys()
        ) else ''
    })

os.makedirs('../Results', exist_ok=True)
with open('../Results/heterogeneity_analysis.csv', 'w', newline='', encoding='utf-8') as f:
    writer = csv.DictWriter(f, fieldnames=output_rows[0].keys())
    writer.writeheader()
    writer.writerows(output_rows)

print("\nResults saved to ../Results/heterogeneity_analysis.csv")


print("\nPreparing text for embedding...")


unique_pairs = {}
for r in rows:
    name = r['OriginalTraitName'].strip()
    defn = r['OriginalTraitDef'].strip()
    unit = r['OriginalTraitUnit'].strip()
    key = (name, defn)
    if key not in unique_pairs:
        unique_pairs[key] = unit

texts = []
pair_list = []
for (name, defn), unit in unique_pairs.items():
    text = f"Trait: {name}. Definition: {defn}. Unit: {unit}."
    texts.append(text)
    pair_list.append((name, defn, unit))

print(f"Total unique pairs to embed: {len(texts)}")
print(f"Example: {texts[0]}")

from sentence_transformers import SentenceTransformer

print("\nLoading model...")
model = SentenceTransformer('all-MiniLM-L6-v2')

print("Generating embeddings...")
embeddings = model.encode(texts, show_progress_bar=True)

print(f"Embeddings shape: {embeddings.shape}")


import Code.clustering_dbcv as clustering_dbcv
import numpy as np

print("\nClustering...")
clusterer = clustering_dbcv.HDBSCAN(
    min_cluster_size=5,
    min_samples=3,
    metric='euclidean'
)
cluster_labels = clusterer.fit_predict(embeddings)

n_clusters = len(set(cluster_labels)) - (1 if -1 in cluster_labels else 0)
n_noise = list(cluster_labels).count(-1)

print(f"Number of clusters: {n_clusters}")
print(f"Noise points (unclustered): {n_noise}")
print(f"Clustered points: {len(cluster_labels) - n_noise}")


from collections import defaultdict

cluster_contents = defaultdict(list)
for i, label in enumerate(cluster_labels):
    cluster_contents[label].append(pair_list[i])

print("\nCluster contents:")
for cluster_id in sorted(cluster_contents.keys()):
    if cluster_id == -1:
        continue
    members = cluster_contents[cluster_id]
    print(f"\n--- Cluster {cluster_id} ({len(members)} members) ---")
    for name, defn, unit in members:
        print(f"  '{name}' | {defn[:50]} | {unit}")


with open('../Results/clustering_results.txt', 'w', encoding='utf-8') as f:
    for cluster_id in sorted(cluster_contents.keys()):
        if cluster_id == -1:
            continue
        members = cluster_contents[cluster_id]
        f.write(f"\n--- Cluster {cluster_id} ({len(members)} members) ---\n")
        for name, defn, unit in members:
            f.write(f"  '{name}' | {defn[:80]} | {unit}\n")

    # noise points
    f.write(f"\n--- NOISE (unclustered, {len(cluster_contents[-1])} items) ---\n")
    for name, defn, unit in cluster_contents[-1]:
        f.write(f"  '{name}' | {defn[:80]} | {unit}\n")

print("Saved to ../Results/clustering_results.txt")


import csv

with open('../Results/clustering_results.csv', 'w', newline='', encoding='utf-8') as f:
    writer = csv.writer(f)
    writer.writerow(['cluster_id', 'n_members', 'OriginalTraitName',
                     'OriginalTraitDef', 'OriginalTraitUnit'])
    for cluster_id in sorted(cluster_contents.keys()):
        members = cluster_contents[cluster_id]
        for name, defn, unit in members:
            writer.writerow([cluster_id, len(members), name, defn, unit])

print("Saved to ../Results/clustering_results.csv")